src/VisualStudio/IntegrationTest/New.IntegrationTests/InProcess/SolutionExplorerInProcess.cs

```diff
@@ -0,0 +1,560 @@
+
+            var project = await GetProjectAsync(projectName, cancellationToken);
+            var references = ((VSProject)project.Object).References.Cast<Reference>().Where(x => x.SourceProject != null).Select(x => x.Name).ToArray();
+            return references;
+        }
+
+        public async Task AddProjectAsync(string projectName, string projectTemplate, string languageName, CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var projectPath = Path.Combine(await GetDirectoryNameAsync(cancellationToken), projectName);
+            var projectTemplatePath = await GetProjectTemplatePathAsync(projectTemplate, ConvertLanguageName(languageName), cancellationToken);
+            var solution = await GetRequiredGlobalServiceAsync<SVsSolution, IVsSolution6>(cancellationToken);
+            ErrorHandler.ThrowOnFailure(solution.AddNewProjectFromTemplate(projectTemplatePath, null, null, projectPath, projectName, null, out _));
+        }
+
+        public async Task RestoreNuGetPackagesAsync(CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var dte = await GetRequiredGlobalServiceAsync<SDTE, EnvDTE.DTE>(cancellationToken);
+            var solution = (EnvDTE80.Solution2)dte.Solution;
+            foreach (var project in solution.Projects.OfType<EnvDTE.Project>())
+            {
+                await RestoreNuGetPackagesAsync(project.FullName, cancellationToken);
+            }
+        }
+
+        public async Task RestoreNuGetPackagesAsync(string projectName, CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var operationProgressStatus = await GetRequiredGlobalServiceAsync<SVsOperationProgress, IVsOperationProgressStatusService>(cancellationToken);
+            var stageStatus = operationProgressStatus.GetStageStatus(CommonOperationProgressStageIds.Intellisense);
+            await stageStatus.WaitForCompletionAsync().WithCancellation(cancellationToken);
+
+            var solutionRestoreService = await GetComponentModelServiceAsync<IVsSolutionRestoreService>(cancellationToken);
+            await solutionRestoreService.CurrentRestoreOperation;
+
+            var projectFullPath = (await GetProjectAsync(projectName, cancellationToken)).FullName;
+            var solutionRestoreStatusProvider = await GetComponentModelServiceAsync<IVsSolutionRestoreStatusProvider>(cancellationToken);
+            if (await solutionRestoreStatusProvider.IsRestoreCompleteAsync(cancellationToken))
+            {
+                return;
+            }
+
+            var solutionRestoreService2 = (IVsSolutionRestoreService2)solutionRestoreService;
+            await solutionRestoreService2.NominateProjectAsync(projectFullPath, cancellationToken);
+
+            while (true)
```

Comment:
could potentially use the retry helper here

Reply:
⏱️ Yep that works